File: web/src/api/modules/task.ts
import service from '@/utils/request'
import type { ApiResponse } from '@/types'

export interface Task {
  id: string
  taskId: string
  taskType: string
  type: string
  expression: string
  description: string
  group: string
  status: string
  statusLabel: string
  createdAt: string
  lastRunAt: string
  nextRunAt: string
  retryCount: number
  maxRetries: number
  concurrency?: string
  timeout?: number
  assignedNode?: string
  nodeGroup?: string
}

export interface TaskLog {
  id: string
  taskId: string
  taskName: string
  group: string
  executeAt: string
  duration: number
  result: string
  errorInfo: string
  retryCount: number
  cronExpr: string
  triggerType: string
  operatorId: string
}

export interface TaskStatistics {
  taskId: string
  taskName: string
  group: string
  totalExecutes: number
  successCount: number
  failedCount: number
  successRate: number
  avgDuration: number
  lastExecuteAt: string
  nextExecuteAt: string
}

export const getTaskList = (params?: {
  page?: number
  pageSize?: number
  taskId?: string
  taskType?: string
}): Promise<ApiResponse<{ list: Task[]; total: number }>> => {
  return service({
    url: '/task',
    method: 'get',
    params,
  })
}

export const getAllTasks = (): Promise<ApiResponse<Task[]>> => {
  return service({
    url: '/task',
    method: 'get',
  })
}

export const getTaskById = (id: string): Promise<ApiResponse<Task>> => {
  return service({
    url: `/task/${id}`,
    method: 'get',
  })
}

export const createTask = (data: {
  taskId: string
  taskType: string
  type: string
  expression: string
  description: string
  group?: string
  retryCount: number
  params: Record<string, unknown>
}): Promise<ApiResponse<Task>> => {
  return service({
    url: '/task',
    method: 'post',
    data,
  })
}

export const updateTask = (
  id: string,
  data: {
    taskId: string
    taskType: string
    type: string
    expression: string
    description: string
    group?: string
    retryCount: number
    params: Record<string, unknown>
  },
): Promise<ApiResponse<Task>> => {
  return service({
    url: `/task/${id}`,
    method: 'put',
    data,
  })
}

export const deleteTask = (id: string): Promise<ApiResponse> => {
  return service({
    url: `/task/${id}`,
    method: 'delete',
  })
}

export const enableTask = (id: string): Promise<ApiResponse> => {
  return service({
    url: `/task/${id}/enable`,
    method: 'post',
  })
}

export const disableTask = (id: string): Promise<ApiResponse> => {
  return service({
    url: `/task/${id}/disable`,
    method: 'post',
  })
}

export const runTask = (id: string): Promise<ApiResponse> => {
  return service({
    url: `/task/${id}/start`,
    method: 'post',
  })
}

export const getTaskLogs = (
  id: string,
  params?: {
    page?: number
    pageSize?: number
  },
): Promise<ApiResponse<{ logs: TaskLog[]; total: number }>> => {
  return service({
    url: `/task/${id}/execution-logs`,
    method: 'get',
    params,
  })
}

export const searchTaskLogs = (params?: {
  taskId?: string
  result?: string
  startDate?: string
  endDate?: string
  page?: number
  pageSize?: number
}): Promise<ApiResponse<{ logs: TaskLog[]; total: number }>> => {
  return service({
    url: '/task/execution-logs',
    method: 'get',
    params,
  })
}

export const startTask = (id: string): Promise<ApiResponse> => {
  return service({
    url: `/task/${id}/start`,
    method: 'post',
  })
}

export const stopTask = (id: string): Promise<ApiResponse> => {
  return service({
    url: `/task/${id}/stop`,
    method: 'post',
  })
}

export const retryTask = (id: string): Promise<ApiResponse> => {
  return service({
    url: `/task/${id}/retry`,
    method: 'post',
  })
}

export const getTaskStatistics = (): Promise<ApiResponse<TaskStatistics[]>> => {
  return service({
    url: '/task/statistics',
    method: 'get',
  })
}

export const setTaskGroup = (id: string, group: string): Promise<ApiResponse> => {
  return service({
    url: `/task/${id}/group`,
    method: 'post',
    data: { group },
  })
}

// 获取任务统计面板数据
export const getTaskDashboard = (): Promise<
  ApiResponse<{
    totalTasks: number
    enabledTasks: number
    disabledTasks: number
    todayExecutions: number
  }>
> => {
  return service({
    url: '/task/dashboard',
    method: 'get',
  })
}

File: server/internal/dto/base/response/task.go
package response

// TaskResponse 任务响应 DTO
type TaskResponse struct {
	ID          int64  `json:"id,string"`
	TaskID      string `json:"taskId"`
	TaskType    string `json:"taskType"`
	Type        string `json:"type"`
	Expression  string `json:"expression"`
	Description string `json:"description"`
	Group       string `json:"group"`
	Status      string `json:"status"`
	StatusLabel string `json:"statusLabel"`
	CreatedAt   string `json:"createdAt"`
	LastRunAt   string `json:"lastRunAt,omitempty"`
	NextRunAt   string `json:"nextRunAt,omitempty"`
	RetryCount  int    `json:"retryCount"`
	MaxRetries  int    `json:"maxRetries"`
	Concurrency string `json:"concurrency"`
	Timeout     int    `json:"timeout"`
}

// TaskListResponse 任务列表响应 DTO
type TaskListResponse struct {
	List  []TaskResponse `json:"list"`
	Total int64          `json:"total"`
}

// TaskExecutionLogResponse 任务执行日志响应 DTO
type TaskExecutionLogResponse struct {
	ID          int64  `json:"id,string"`
	TaskID      int64  `json:"taskId,string"`
	TaskName    string `json:"taskName"`
	Group       string `json:"group"`
	ExecuteAt   string `json:"executeAt"`
	Duration    int64  `json:"duration"`
	Result      string `json:"result"`
	ErrorInfo   string `json:"errorInfo"`
	RetryCount  int    `json:"retryCount"`
	CronExpr    string `json:"cronExpr"`
	TriggerType string `json:"triggerType"`
	OperatorID  int64  `json:"operatorId,string"`
}

// TaskExecutionLogsResponse 任务执行日志列表响应 DTO
type TaskExecutionLogsResponse struct {
	Logs  []TaskExecutionLogResponse `json:"logs"`
	Total int64                      `json:"total"`
}

// TaskStatisticsResponse 任务统计响应 DTO
type TaskStatisticsResponse struct {
	TaskID        int64   `json:"taskId,string"`
	TaskName      string  `json:"taskName"`
	Group         string  `json:"group"`
	TotalExecutes int64   `json:"totalExecutes"`
	SuccessCount  int64   `json:"successCount"`
	FailedCount   int64   `json:"failedCount"`
	SuccessRate   float64 `json:"successRate"`
	AvgDuration   float64 `json:"avgDuration"`
	LastExecuteAt string  `json:"lastExecuteAt"`
	NextExecuteAt string  `json:"nextExecuteAt"`
}

// TaskStatisticsListResponse 任务统计列表响应 DTO
type TaskStatisticsListResponse struct {
	List []TaskStatisticsResponse `json:"list"`
}
